src/pip/_internal/resolution/resolvelib/provider.py

```diff
@@ -3,26 +3,101 @@
+# ideas. The API of Requirement and Candidate objects are defined in the base
+# classes, but essentially map fairly directly to the equivalent provider
+# methods. In particular, `find_matches` and `is_satisfied_by` are
+# requirement methods, and `get_dependencies` is a candidate method.
+#
+# The factory is the interface to pip's internal mechanisms. It is stateless,
+# and is created by the resolver and held as a property of the provider. It is
+# responsible for creating Requirement and Candidate objects, and provides
+# services to those objects (access to pip's finder and preparer).
+
 
 class PipProvider(AbstractProvider):
     def __init__(
         self,
         factory,  # type: Factory
         constraints,  # type: Dict[str, SpecifierSet]
         ignore_dependencies,  # type: bool
+        upgrade_strategy,  # type: str
+        roots,  # type: Set[str]
     ):
         # type: (...) -> None
         self._factory = factory
         self._constraints = constraints
         self._ignore_dependencies = ignore_dependencies
+        self._upgrade_strategy = upgrade_strategy
+        self.roots = roots
+
+    def sort_matches(self, matches):
+        # type: (Sequence[Candidate]) -> Sequence[Candidate]
+
+        # The requirement is responsible for returning a sequence of potential
+        # candidates, one per version. The provider handles the logic of
+        # deciding the order in which these candidates should be passed to
+        # the resolver.
+
+        # The `matches` argument is a sequence of candidates, one per version,
+        # which are potential options to be installed. The requirement will
+        # have already sorted out whether to give us an already-installed
+        # candidate or a version from PyPI (i.e., it will deal with options
+        # like --force-reinstall and --ignore-installed).
+
+        # We now work out the correct order.
+        #
+        # 1. If no other considerations apply, later versions take priority.
+        # 2. An already installed distribution is preferred over any other,
+        #    unless the user has requested an upgrade.
+        #    Upgrades are allowed when:
+        #    * The --upgrade flag is set, and
+        #      - The project was specified on the command line, or
+        #      - The project is a dependency and the "eager" upgrade strategy
+        #        was requested.
+
+        def _eligible_for_upgrade(name):
+            # type: (str) -> bool
+            if self._upgrade_strategy == "eager":
+                return True
+            elif self._upgrade_strategy == "only-if-needed":
+                return (name in self.roots)
+            return False
+
+        def keep_installed(c):
+            # type: (Candidate) -> int
+            """Give priority to an installed version?"""
+            if not is_already_installed(c):
+                return 0
+
+            if _eligible_for_upgrade(c.name):
+                return 0
+
+            return 1
```

Comment:
```suggestion
            """Give priority to an installed version?"""
            return (
                is_already_installed(c) and
                upgrade_permitted_by_upgrade_strategy(c.name)
            )
```

Knowing True == 1 and False == 0, this is *much* more readable IMO.

Reply:
IIRC, this is the kind of stuff for which bool is a subclass of int, so I'd actually be a little bummed if we don't do this. :)

Reply:
I pondered about doing something but eventually decided I favour using int slightly. I need a microsecond to process boolean as an orderable value.

Reply:
BUT it reads so well once you figure that out!

Reply:
I'm with @uranusjr, in my head I just end up converting the boolean logic back into a conditional, so why not have the conditionals? Also, why did that "not" appear in there? That confuses me as well, suggesting that it would make it harder to maintain, as well.

I do think it's excessively split up - two sub-functions for this one bit of logic - but I prefer "if it was hard to write, it should be easy to read" over "it it was hard to write, it should be hard to read". So sue me 🙂

Reply:
> Also, why did that "not" appear in there?

Wait, isn't `not _eligible_for_upgrade(c.name)` the same as the following?

```
            if _eligible_for_upgrade(c.name):
                return 0
            return 1
```

Reply:
Yes. That's me demonstrating why even mildly complex booleans aren't easy for me to reason about 😉 Is the updated version I pushed a little while ago any more to your liking?